### ISSUE TITLE
[java] UnusedX rules should ignore some variable names and annotations

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2838]
**Is your feature request related to a problem? Please describe.**
[UnusedLocalVariable](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedlocalvariable), [UnusedFormalParameter](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedformalparameter) flag all unused variables.

**Describe the solution you'd like**
Please  add a property  to the rule specifying a pattern such as starts with `$`,  `_` or `unused` for the variable name so that variables that meet the criteria are excluded. Additionally, you could support custom annotations that mark the variables as Unused. e.g., `@Unused`.

**Describe alternatives you've considered**
Example code flagged by [ForLoopCanBeForEach](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#forloopcanbeforeach) when usual for loop used:

```java
@SuppressWarnings("PMD.UnusedLocalVariable")
  @Override
  public Solution solve() {
    int[] dp = new int[capacity + 1];

    List<Item> itemsList = new ArrayList<>(items.length);
    int i = 0;
    for (int dpVal : dp) {
      for (Item item : items) {
        if (item.weight <= i) {
          int includedVal = dp[i - item.weight] + item.value;
          if (includedVal > dp[i]) dp[i] = includedVal;
        }
      }
      ++i;
    }
    path(items, capacity, dp, itemsList);
    itemsList = Item.pack(itemsList);
    return new Solution(itemsList, dp[capacity]);
  }
```

**Additional context**
[Haskell/Python style](https://stackoverflow.com/a/32862238/3924108)
[Error-prone](https://errorprone.info/bugpattern/UnusedVariable)

Also see #2923

**Comment by Fernal73** [comment]
This can, however, mess with your [LocalVariableNamingConventions] (https://pmd.github.io/latest/pmd_rules_java_codestyle.html#localvariablenamingconventions) rule.

Easier to suppress the violation?

**Comment by Fernal73** [comment]
Another option might be to prefix the variable name with `unused`.
Borrowed from [Error-Prone(https://errorprone.info/bugpattern/UnusedVariable)

**Comment by oowekyala** [comment]
[UnusedAssignment](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedassignment) recognises the prefix `ignored`:

https://github.com/pmd/pmd/blob/40a8ac6f5cbf47b457eef863021371eb42fdbcc5/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnusedAssignmentRule.java#L220-L223

Both these rules should support the same set of prefixes. We could add `unused` to this set and share the method between those rules

**Comment by Fernal73** [comment]
You have a point since the same property holds for multiple rules. Updating them to keep them consistent with each other will be a pain. Until PMD supports common properties that can be shared across rules, hard-coding the prefixes is the way to go. The documentation, however, will have to be up-front about these exceptions; you can't expect users to wade through every one of the code examples supplied to discover how to contravene the rule(s).

**Comment by oowekyala** [comment]
>  Until PMD supports common properties that can be shared across rules, hard-coding the prefixes is the way to go.

I don't think there's really a need to support configuring this though. The ErrorProne check you linked doesn't support configuring this, nor do the Intellij or Eclipse inspections. IMHO, programmers should strive to use the same conventions across all projects. I think "ignored" or "unused" are explicit, somewhat conventional for exceptions at least, and it makes us compatible with ErrorProne and Intellij. Besides, when a variable is ignored, you preferably don't want to bother giving it a name (which is why `_` is commonly used across languages for this purpose). Prefixing a name with `_` or `$` doesn't avoid this and is less clear, not to mention that it may violate naming convention rules.

Users can anyway use violation suppression to add exceptions to the rule.

**Comment by Fernal73** [comment]
I default to not being comfortable with hard-coding values. In this case,
there is no easy alternative.

 Else users need to modify the NamingConvention regexes to allow for
variables beginning with `_` or `$`. That's  advanced; you wouldn't be
messing with default configurations unless you understand what you're doing.

Lombok variables begin with `$` but allow for violation suppressions by
optionally generating a `@SuppressWarnings("all")` for  its methods and
fields. FYI.

**Comment by Fernal73** [comment]
> > Until PMD supports common properties that can be shared across rules, hard-coding the prefixes is the way to go.
> 
> I don't think there's really a need to support configuring this though. The ErrorProne check you linked doesn't support configuring this, nor do the Intellij or Eclipse inspections. IMHO, programmers should strive to use the same conventions across all projects. I think "ignored" or "unused" are explicit, somewhat conventional for exceptions at least, and it makes us compatible with ErrorProne and Intellij. Besides, when a variable is ignored, you preferably don't want to bother giving it a name (which is why `_` is commonly used across languages for this purpose). Prefixing a name with `_` or `$` doesn't avoid this and is less clear, not to mention that it may violate naming convention rules.
> 
> Users can anyway use violation suppression to add exceptions to the rule.

While I appreciate the need to be compatible with Intellij and/or Error-Prone and not be in violation of default Naming Conventions, I , as a programmer , would like anyone reading my code to assign special significance to an unused variable. This can be best done by prefixing its name with `_` or `$`preferable `$ignored` or `_unused` or any other combination. I would then simply suppress any naming convention warnings assuming I don't wish to go to the trouble of updating the existing  regexes.

**Comment by Fernal73** [comment]
> I default to not being comfortable with hard-coding values. In this case, there is no easy alternative. Else users need to modify the NamingConvention regexes to allow for variables beginning with `_` or `$`. That's advanced; you wouldn't be messing with default configurations unless you understand what you're doing. Lombok variables begin with `$` but allow for violation suppressions by optionally generating a `@SuppressWarnings("all")` for its methods and fields. FYI.

My preference for properties stems from the fact that if values are embedded in the code, then the earliest any change request can be accommodated is the next release. Modifiable configurations, on the other hand, allows for some changes without having to raise an issue.

**Comment by Fernal73** [comment]
After all the discussions related to annotations elsewhere,  I think these rules should have a common property that lists a set of annotations that mark parameters and variables to be ignored by these rules.

**Comment by oowekyala** [comment]
> After all the discussions related to annotations elsewhere, I think these rules should have a common property that lists a set of custom annotations that mark parameters and variables to be ignored by these rules.

> I must admit that I haven't come across any examples of annotating unused variables or parameters.

I think it's quite rich that you're asking for a change for which you have no use case (and it's not the first time). This rule exists since 2002 and AFAIK nobody has asked about support for custom suppression annotations. Not being too configurable promotes conventions over idiosyncrasy, which I believe is a good thing. It also makes PMD simpler to use.

> I , as a programmer , would like anyone reading my code to assign special significance to an unused variable. This can be best done by prefixing its name with _ or $preferably $ignored, _unused, $unused or _ignored. I would then simply suppress any naming convention warnings assuming I don't wish to go to the trouble of updating the existing regexes.

You can already do this with suppression... Why would you be ok with suppressing the naming conventions warnings but not the unused ones...

**Comment by Fernal73** [comment]
> 
> > I must admit that I haven't come across any examples of annotating unused variables or parameters.

Just because there are no examples doesn't mean there's no usecase. The Scala example proves there is, IMO.

> 
>
> 
> > I , as a programmer , would like anyone reading my code to assign special significance to an unused variable. This can be best done by prefixing its name with _ or $preferably $ignored, _unused, $unused or _ignored. I would then simply suppress any naming convention warnings assuming I don't wish to go to the trouble of updating the existing regexes.
> 
> 

I can always update the naming conventions regex if needed.

**Comment by Fernal73** [comment]
> > I must admit that I haven't come across any examples of annotating unused variables or parameters.
> 

I made this statement based on Google search results. I don't think trawling through GitHub is going to turn up any relevant results that easily. They may exist. I don't know. Do you?

**Comment by oowekyala** [comment]
> Also, I think having to rename variables to circumvent a rule sucks.

Isn't this exactly what this ticket is about?

> I made this statement based on Google search results. I don't think trawling through GitHub is going to turn up any relevant results that easily. They may exist. I don't know. Do you?

I just think that features for which users "may exist" are dubious candidates for feature requests, especially when there are arguments for not implementing them. It's especially aggravating when it's reported as an addendum to a ticket which is essentially about something else. You do this quite often and it's getting old.

**Comment by Fernal73** [comment]
> I just think that features for which users "may exist" are dubious
> candidates for feature requests, especially when there are arguments for
> not implementing them. It's especially aggravating when it's reported as an
> addendum to a ticket which is essentially about something else. You do this
> quite often and it's getting old.
>

There has been no movement on this ticket as yet. And the issue can always
be rewritten to include this feature. All change is aggravating. I think
this is relevant to the original issue raised, so it should be commented on
here.
If you wish to fix it separately, it can be written up as a separate issue.
But it's still connected to the original issue.

>

**Comment by Fernal73** [comment]
>>You do this quite often and it's getting old.

You're right. I do do that quite often and it must seem repetitive.
So what you're suggesting instead is that I not add anything to current issues unless it's what's already laid out and create a separate issue instead. Also, then I not create issues when they're not fully-formed or as perfect or complete that I'd expect it to be. In other words, not wait for any reaction to fully flesh it out. 
And you and the PMD team will decide whether two or more issues are related and work your Merge magic.

**Comment by Fernal73** [comment]
> > All unexpected change is aggravating when it involves additional work.
> 
> I'm telling you there's a problem with the way you interact with this project; don't dismiss it as just laziness on my part, it's patronizing
> 
> > But it's still connected to the original issue.
> > And the issue can always be rewritten to include this feature.
> 
> Do you not see a problem with this? Editing issues beyond recognition breaks cross-referencing, makes the search engine less useful, and makes it difficult for us maintainers to just keep mental tabs on them. Related issues can _reference_ each other. If they can be solved together, we can merge them. But the initial discussion should ideally be focused into its own ticket and not be edited 10 times. Your tendency to continually expand the scope of existing issues just makes them more difficult to solve.
> 
> In case you haven't noticed, in this specific ticket I'm ok to implement the part about ignoring names, but more reserved for custom annotations. The two points can be discussed and solved independently. So the way you edited your post is exactly the problem I'm talking about: you've taken something that initially had a single action item and could have been solved easily, and tangled it with something else, that will either have to be split up into another issue, or be ignored.
> 
> > So what you're suggesting instead is that I not add anything to current issues unless it's what's already laid out and create a separate issue instead
> 
> Why is that a bad thing? I mean seriously, do you think that [random comments like this](https://github.com/pmd/pmd/issues/2943#issuecomment-734730248) are helping the original ticket or the new point you're making in any way? Do you not think that creating a new ticket would actually serve your point better, and also make it worth a discussion, because you'd have to pause to motivate your thought?
> 
> Maybe a good way to think about it is that each issue should be a _single_ discussion. If you want to add a comment that you know could start another discussion entirely, do it in a new issue.

Without the annotations, it is not a complete solution. Period.
